### PR TITLE
nixos: xserver.displayManager: use slim for automatic logins

### PR DESCRIPTION
**Pull request by oxij**
###### Motivation for this change

`lightdm` is not that light for automatic logins: it runs two processes three threads each (while slim runs one of each) all of which stay in memory even after login is finished. Meanwhile, `auto` users usually want the display manager to get out of their way, not to pollute their system with six useless threads.  Moreover, the closure size of slim+theme in its default config in NixOS is 53M, while the similar closure size of lightdm+greeter+theme is 254M.

###### Things done

- [X] This is how it was before #30890.
- [X] Writing this from a system with this applied.

### CODE REVIEW

**Comment by xeji** [comment]
If this is merged (I don't have a strong opinion about it), it should be backported to 18.09 because it can be a pain to debug/fix NixOS tests if the basic test setup is different between master and 18.09.

File: nixos/modules/services/x11/display-managers/auto.nix [hunk 1]
```diff
@@ -41,28 +41,12 @@ in
...
 
-    services.xserver.displayManager.lightdm = {
+    services.xserver.displayManager.slim = {
       enable = true;
-      autoLogin = {
-        enable = true;
-        user = cfg.user;
-      };
+      autoLogin = true;
+      defaultUser = cfg.user;
```
Comment (matthewbauer) [inline nixos/modules/services/x11/display-managers/auto.nix]: Not an issue with this PR but we should standardize the settings so ```autoLogin``` works the same regardless of DM.

**Comment by oxij** [comment]
Ping. I still think this need to be merged.

Let me remind you that this also fixes

> - Normally I `exec xmonad` at the end of `.xsession`. With `lightdm` exiting `xmonad` brings the system into an unusable state, `lightdm` just keeps running, showing its background window as if nothing happened, and you have to login in another tty and restart `display-manager` manually to get anything done.

mentioned in https://github.com/NixOS/nixpkgs/pull/30890#issuecomment-419670755

Also of note since the switch to `lightdm` are the following discourse topics
https://discourse.nixos.org/t/stuck-with-a-failing-lightdm/930/1 and
https://discourse.nixos.org/t/how-to-upgrade-from-18-03-to-18-09/933/1

Both have "`lightdm` doesn't works until you rebuild and reboot your system several times" symptoms, no idea why or what, but looks interesting.

**Comment by wizeman** [comment]
This PR also fixes another bug: if the desktop manager crashes, it causes lightdm to restart but it won't auto-login again, therefore getting stuck at the 'enter password' screen.
I was observing this behavior with Kodi, but when I tried applying this PR to my config, the bug is fixed.

**Comment by samueldr** [comment]
The "bug" about lightdm not doing auto-login on crash (haven't checked, but I suspect it's when the session exits non-zero) is imo not a bug. This reduces the risk of a broken session causing a tight loop with the display manager `logging-in → starting session → crashing → restarting`, in those situations switching to a VT could be problematic since AFAIK most DMs will throw away the X server and start a new one, thus pulling the user back to the looping VT.

I do not know how to approach this appropriately, but maybe some tip in the documentation that for unattended autologin slim would be more appropriate due to that behaviour, while the behaviour from lightdm is preferred on an end-user's machine.

**Comment by wizeman** [comment]
@samueldr IMO your scenario is more appropriately handled by lightdm exiting with a non-zero status (assuming the DM exited with a non-zero status) and letting systemd auto-restart the `display-manager` service. This would probably work better because systemd already has in-built mechanisms to avoid a service from entering an infinite start-fail loop (after a few tries, systemd stops trying to restart the service when the service fails relatively quickly after being restarted).

This might be what slim does already, although I haven't verified it yet.

**Comment by wizeman** [comment]
Ok, I have verified that `slim` behaves exactly as I mentioned in my comment above, which is the behavior I would expect and prevents the infinite start-fail loop in case the desktop manager keeps crashing at startup.

To check this behavior, I have performed a bunch of `pkill -9 kodi-x11` in a row (`kodi-x11` is the process name of the desktop manager I am using) on a machine with this PR, and I confirm that `systemd` stopped restarting the `display-manager` service (i.e. the X server) after I killed kodi 3 times in less than one minute.

Note that `systemd` already does this by default, but the "stop restarting" behavior is also explicitly fine-tuned here:
https://github.com/NixOS/nixpkgs/blob/3223534f89a3be5e539e329b3b7a5bc3593a94e8/nixos/modules/services/x11/xserver.nix#L684-L687

To be clear, note that `lightdm` currently does not behave like this -- instead, if the desktop manager crashes `lightdm` will simply go back to a login screen even though auto-login is enabled, which is not the behavior I would expect (principle of least surprise?).

So to summarize, IMHO I think this PR fixes a bunch of issues with auto-login and I think it should either be merged or `lightdm` should be fixed accordingly.

**Comment by samueldr** [comment]
Looks like lightdm [has an open issue](https://bugs.launchpad.net/lightdm/+bug/1396824).

**Comment by fgaz** [comment]
Slim was removed in #73251, I think this can be closed

**Comment by oxij** [comment]
No, the correct way now is to revert #73251 and then merge this. The OP issue still persists.
